README
##
##  README - What is this all about...
##  Copyright (c) 2000 Yeak Nai Siew, All Rights Reserved. 
##

##############################################################################
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

##############################################################################

CHANGES
-------
v0.1	24 Apr 2000, Initial public release.
v0.2	22 Mar 2004, added handling for certificate renewal
v0.3    08 Mar 2016, see CHANGELOG

This README file will explain how to use this handy "kit" to get your
certificate creation and signing task easier using openssl. You can also
modify the scripts to suit your needs.

This scripts are particularly designed to handle the web server cert
generation and user cert generation (S/MIME). Then the script let you
package the certs into pkcs12 format for you to download to your Netscape
Communicator to use.

Most of the codes are borrowed from RSE's Mod-SSL package. I only "extend"
it and probably make it easier to use.


Introduction
------------

It took me 2 full days to be able to come out this kit, and months of
past experience to understand OpenSSL's operation. I hope this kit can help
anyone to get things work, up and run immediately. If this helped you,
please let me know. If this does not help you, let me know so that I can
improve it.

I use this kit to generate certificates for school and friends to use. It
happend that recently I only have one Solaris 8 and need to setup
a secured web-based authentication that used certificate + conventional
login process. So, the best I can do now is this:

1. Use Apache as core Web Server
2. Use OpenSSL as core SSL encryption
3. Use Mod-SSL to glue Apache and OpenSSL
4. Use PHP3 with LDAP capability to do my programming task
5. Use Netscape Directory Server 4.11 for backend database purpose (can be
   PostgreSQL or MySQL)

I won't go into the details of compiling Apache, OpenSSL, Mod-SSL, PHP3
and LDAPSDK things... the documentation from Apache and Mod-SSL is
particularly useful. Great job has been done by RSE many months ago!


Requirement to run this program
-------------------------------

1. Unix system that can run "sh". I prefer Linux. But I write this on
   Solaris 8 due to work requirement. Yes, it runs on Linux.

2. Working openssl 0.9.4 or above. I am using 0.95a. Yes, it runs on Linux.

To test the certification, you need,

1. Web Server that use the server certificate

2. Netscape Communicator 4.5 and above that use the S/MIME certificate.
   I have tested IE5 on Windows 2000. Also the Outlook Express. Work.


How to use it
-------------

The script is written based on this concept:

1. You have one self-signed Root Certificate (Root CA) that is used to
   sign everything that you are going to generate.
2. You can generate the web server cert for Apache+Mod-SSL to use.
3. You can generate S/MIME key for your Netscape browser to use.

To start, you must create a self-signed Root Certification Authority.
This is done by running "new-root-ca.sh". Remember the password of this
key.

You run "new-server-cert.sh" to create a web server's private and
public key. The script will continue on to enter the ceritificate information.
We call this "certificate signing request (csr)". The format is x509. You
need to run "sign-server-cert.sh" to approve and sign with your root key.

You run "new-user-cert.sh" to create a user certificate. You can use this
to sign and encrypt your email using MS Outlook Express or Netscape browser.
You need to run "sign-user-cert.sh" to sign it. Finally,  you run "p12.sh"
to package the private key, signed key, together with CA's Public key into
a file with ".p12" extension. You can download this to your PC and import in.

I hope the explaination is very straight forward.

Example Usage
-------------


1. Create the self-signed Root CA key


# ./new-root-ca.sh 
No Root CA key round. Generating one
Generating RSA private key, 1024 bit long modulus
....++++++
.............++++++
e is 65537 (0x10001)
Enter PEM pass phrase:
Verifying password - Enter PEM pass phrase:

Self-sign the root CA...
Using configuration from root-ca.conf
Enter PEM pass phrase:
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [MY]:
State or Province Name (full name) [Perak]:
Locality Name (eg, city) [Sitiawan]:
Organization Name (eg, company) [My Directory Sdn Bhd]:
Organizational Unit Name (eg, section) [Certification Services Division]:
Common Name (eg, MD Root CA) []:MD Root CA
Email Address []:md-ca@md.com.my



2. Create the server cert


# ./new-server-cert.sh 
Usage: ./new-server-cert.sh <www.domain.com>
# ./new-server-cert.sh www.md.com.my
No www.md.com.my.key round. Generating one
Generating RSA private key, 1024 bit long modulus
...++++++
......++++++
e is 65537 (0x10001)

Fill in certificate data
Using configuration from server-cert.conf
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [MY]:
State or Province Name (full name) [Perak]:    
Locality Name (eg, city) [Sitiawan]:
Organization Name (eg, company) [My Directory Sdn Bhd]:
Organizational Unit Name (eg, section) [Secure Web Server]:
Common Name (eg, www.domain.com) []:www.md.com.my
Email Address []:info@md.com.my

You may now run ./sign-server-cert.sh to get it signed



3. Sign the server cert


# ./sign-server-cert.sh 
Usage: ./sign-server-cert.sh <www.domain.com>
# ./sign-server-cert.sh www.md.com.my
CA signing: www.md.com.my.csr -> www.md.com.my.crt:
Using configuration from ca.config
Enter PEM pass phrase:
Check that the request matches the signature
Signature ok
The Subjects Distinguished Name is as follows
countryName           :PRINTABLE:'MY'
stateOrProvinceName   :PRINTABLE:'Perak'
localityName          :PRINTABLE:'Sitiawan'
organizationName      :PRINTABLE:'My Directory Sdn Bhd'
organizationalUnitName:PRINTABLE:'Secure Web Server'
commonName            :PRINTABLE:'www.md.com.my'
emailAddress          :IA5STRING:'info@md.com.my'
Certificate is to be certified until Apr 24 12:43:27 2001 GMT (365 days)
Sign the certificate? [y/n]:y


1 out of 1 certificate requests certified, commit? [y/n]y
Write out database with 1 new entries
Data Base Updated
CA verifying: www.md.com.my.crt <-> CA cert
www.md.com.my.crt: OK



4. Create the user cert


# ./new-user-cert.sh 
Usage: ./new-user-cert.sh user@email.address.com
# ./new-user-cert.sh yeak@md.com.my
No yeak@md.com.my.key round. Generating one
Generating RSA private key, 1024 bit long modulus
....................++++++
............++++++
e is 65537 (0x10001)

Fill in certificate data
Using configuration from user-cert.conf
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Common Name (eg, John Doe) []:Yeak Nai Siew    
Email Address []:yeak@md.com.my

You may now run ./sign-user-cert.sh to get it signed



5. Sign the user cert


# ./sign-user-cert.sh     
Usage: ./sign-user-cert.sh user@email.address.com
# ./sign-user-cert.sh yeak@md.com.my
CA signing: yeak@md.com.my.csr -> yeak@md.com.my.crt:
Using configuration from ca.config
Enter PEM pass phrase:
Check that the request matches the signature
Signature ok
The Subjects Distinguished Name is as follows
commonName            :PRINTABLE:'Yeak Nai Siew'
emailAddress          :IA5STRING:'yeak@md.com.my'
Certificate is to be certified until Apr 24 12:53:58 2001 GMT (365 days)
Sign the certificate? [y/n]:y


1 out of 1 certificate requests certified, commit? [y/n]y
Write out database with 1 new entries
Data Base Updated
CA verifying: yeak@md.com.my.crt <-> CA cert
yeak@md.com.my.crt: OK




6. Package the user certificate into a single pkcs12 file.



# ./p12.sh 
Usage: ./p12.sh user@email.address.com
# ./p12.sh yeak@md.com.my
Enter Export Password:
Verifying password - Enter Export Password:

The certificate for yeak@md.com.my has been collected into a pkcs12 file.
You can download to your browser and import it.



7. DONE!



PROBLEM SOLVING
---------------
I am sure you have problem to even using the script. Yes, I actually need
to improve them first. But see the following first (I know, I know...)

Q. "openssl" not found?

A. You need to have openssl 0.9.4 in your machine. For Red Hat Linux, just
   install the openssl-xxxx.rpm package. On Solaris, make sure your PATH
   environment is set correctly. In my case, I have openssl installed in
   /usr/local/ssl/bin/openssl. So I need to make sure my PATH contain
   "/usr/local/ssl/bin" and export the this PATH variable.

Q. warning, not much extra random data, consider using the -rand option
   14044:error:20.......................... PRNG not seeded:md_rand.c

A. In openssl 0.9.5a version, the random seed (PRNG) is important in order to
   generate any private key. This is a security feature, not a bug. In
   my kit, I actually include the random-bits that I make up. Do not simply
   use it! You should try to modify the "new-root-ca.sh" file, look for
   "-rand random-bits", change it to "-rand /var/adm/wtmpx:/var/adm/messages"
   in order to get your own unique seed. Note, once you successfully make
   the openssl run, it will create a $HOME/.rnd file for future use.

   In Linux, you can use "-rand /dev/urandom". If you use
   "-rand /dev/random", be sure to patiently wait until it start to generate
   something... :-)

Q. where should I install ssl.ca directory?

A. You can put anywhere. I put inside /usr/local/apache/conf where I have
   ssl.key, ssl.crt, ssl.csr... around.

Q. I can't important the p12 files to my Mac or PC!

A. Make sure the file is downloaded as "binary". Windows NT 4 might have
   problem to read the key. Try Windows 2000 or Windows 98.


TODO
----
Web based certificate management. Need a lot of input from all of you out
there.


CONTACT
-------
This is my signature (my vcard!)

Yeak Nai Siew      [NIC:NY628] << Mac OS Forever >> << Linux Forever>>
(yeak@md.com.my | yeak@mac.com)                  http://www.md.com.my/
e-certificate      http://www.brainbench.com/transcript.jsp?pid=120196
ICQ#: 13391181                                Chief Technology Officer



CHANGELOG
-------------------------------------------------------------------------------
Release v0.3, 08-MAR-2016

    Updates from Eric Dey

    - Added handling for Subject Alternate Names as DNS host names and IPs.
    - Parameterized key bits; default 2048.
    - Parameterized hash algorithm; default SHA256.
    - Parameterized certificate expiration days; default 730 for server/user
    - Use /dev/urandom for entropy instead of static "random-bits" file. (1)
    - Changed from DES3 to AES256 to encrypt Root CA key.
    - Server key usage as: nonRepudiation, digitalSignature, keyEncipherment
    - All scripts changed to use BASH.
    - Miscellaneous script clean-ups.


Note 1: See README for a discussion about the entropy source.


-------------------------------------------------------------------------------
Release v0.2, 22-MAR-2004

    Updates by Mike Rhyner to add handling for renewal of certificates.


-------------------------------------------------------------------------------
Release v0.1, 24-APR-2000

    Initial public release from Yeak Nai Siew.


p12.sh
#!/bin/bash
##
##  p12.sh - Collect the user certs and pack into pkcs12 format
##  Copyright (c) 2000 Yeak Nai Siew, All Rights Reserved. 
##

CERT=$1
if [ $# -ne 1 ]; then
        echo "Usage: $0 user@email.address.com"
        exit 1
fi

# Check for requirement
if [ ! -f $CERT.key -o ! -f $CERT.crt -o ! -f ca.crt ]; then
	echo ""
        echo "Cannot proceed because:"
	echo "1. Must have root CA certification"
	echo "2. Must have $CERT.key"
	echo "1. Must have $CERT.crt"
	echo ""
	exit 1
fi

username="`openssl x509 -noout  -in $CERT.crt -subject | sed -e 's;.*CN=;;' -e 's;/Em.*;;'`"
caname="`openssl x509 -noout  -in ca.crt -subject | sed -e 's;.*CN=;;' -e 's;/Em.*;;'`"

# Package it.
openssl pkcs12 \
	-export \
	-in "$CERT.crt" \
	-inkey "$CERT.key" \
	-certfile ca.crt \
	-name "$username" \
	-caname "$caname" \
	-out $CERT.p12

echo ""
echo "The certificate for $CERT has been collected into a pkcs12 file."
echo "You can download to your browser and import it."
echo ""


new-root-ca.sh
#!/bin/bash
##
##  new-root-ca.sh - create the root CA
##  Copyright (c) 2000 Yeak Nai Siew, All Rights Reserved. 
##

KEYBITS=2048
HASHALGO="sha256"
VALID_DAYS=3650
RANDOM_SRC=/dev/urandom

# Create the master CA key. This should be done once.
if [ ! -f ca.key ]; then
	echo "No Root CA key found. Generating one"
	openssl genrsa -aes256 -out ca.key $KEYBITS -rand ${RANDOM_SRC}
	echo ""
fi

# Self-sign it.
CONFIG="root-ca.conf"
cat >$CONFIG <<EOT
[ req ]
default_bits			= $KEYBITS
default_keyfile			= ca.key
default_md              = $HASHALGO
distinguished_name		= req_distinguished_name
x509_extensions			= v3_ca
string_mask			= nombstr
req_extensions			= v3_req
[ req_distinguished_name ]
countryName			= Country Name (2 letter code)
countryName_default		= US
countryName_min			= 2
countryName_max			= 2
stateOrProvinceName		= State or Province Name (full name)
stateOrProvinceName_default	= Texas
localityName			= Locality Name (eg, city)
localityName_default		= Austin
0.organizationName		= Organization Name (eg, company)
0.organizationName_default	= My Personal Certificate Authority
organizationalUnitName		= Organizational Unit Name (eg, section)
organizationalUnitName_default	= Certification Services Division
commonName			= Common Name (eg, MD Root CA)
commonName_max			= 64
emailAddress			= Email Address
emailAddress_max		= 40
[ v3_ca ]
basicConstraints		= critical,CA:true
subjectKeyIdentifier		= hash
[ v3_req ]
nsCertType			= objsign,email,server
EOT

echo "Self-sign the root CA..."
openssl req -new -x509 -days ${VALID_DAYS} -config $CONFIG -key ca.key -out ca.crt

rm -f $CONFIG


new-user-cert.sh
#!/bin/bash
##
##  new-user-cert.sh - create the user cert for personal use.
##  Copyright (c) 2000 Yeak Nai Siew, All Rights Reserved. 
##

KEYBITS=2048
HASHALGO="sha256"

# Create the key. This should be done once per cert.
CERT=$1
if [ $# -ne 1 ]; then
        echo "Usage: $0 user@email.address.com"
        exit 1
fi

# if private key exists, ask if we want to generate a new key
if [ -f $CERT.key ]; then
  read -p "a key for this cn is already existing, generate a new one? " ANSWER
  if [ "$ANSWER" == "Y" ] || [ "$ANSWER" == "y" ]; then
    rm -f $CERT.key
  fi
fi

if [ ! -f $CERT.key ]; then
	echo "No $CERT.key found. Generating one"
	openssl genrsa -out $CERT.key $KEYBITS
	echo ""
fi

# Fill the necessary certificate data
CONFIG="user-cert.conf"
cat >$CONFIG <<EOT
[ req ]
default_bits			= $KEYBITS
default_keyfile			= user.key
default_md              = $HASHALGO
distinguished_name		= req_distinguished_name
string_mask			= nombstr
req_extensions			= v3_req
[ req_distinguished_name ]
commonName			= Common Name (eg, John Doe)
commonName_max			= 64
emailAddress			= Email Address
emailAddress_max		= 40
[ v3_req ]
nsCertType			= client,email
basicConstraints		= critical,CA:false
EOT

echo "Fill in certificate data"
openssl req -new -config $CONFIG -key $CERT.key -out $CERT.csr

rm -f $CONFIG

echo ""
echo "You may now run ./sign-user-cert.sh to get it signed"


sign-user-cert.sh
#!/bin/bash
##
##  sign-user-cert.sh - sign using our root CA the user cert
##  Copyright (c) 2000 Yeak Nai Siew, All Rights Reserved. 
##

HASHALGO="sha256"
VALID_DAYS=730
RANDOM_SRC=/dev/urandom

CERT=$1
if [ $# -ne 1 ]; then
        echo "Usage: $0 user@email.address.com"
        exit 1
fi
if [ ! -f $CERT.csr ]; then
        echo "No $CERT.csr found. You must create that first."
	exit 1
fi
# Check for root CA key
if [ ! -f ca.key -o ! -f ca.crt ]; then
	echo "You must have root CA key generated first."
	exit 1
fi

# Sign it with our CA key #

#   make sure environment exists
if [ ! -d ca.db.certs ]; then
    mkdir ca.db.certs
fi
if [ ! -f ca.db.serial ]; then
    echo '01' >ca.db.serial
fi
if [ ! -f ca.db.index ]; then
    cp /dev/null ca.db.index
fi

#  create the CA requirement to sign the cert
cat >ca.config <<EOT
[ ca ]
default_ca              = default_CA
[ default_CA ]
dir                     = .
certs                   = \$dir
new_certs_dir           = \$dir/ca.db.certs
database                = \$dir/ca.db.index
serial                  = \$dir/ca.db.serial
RANDFILE                = ${RANDOM_SRC}
certificate             = \$dir/ca.crt
private_key             = \$dir/ca.key
default_days            = ${VALID_DAYS}
default_crl_days        = 30
default_md              = $HASHALGO
preserve                = yes
x509_extensions		= user_cert
policy                  = policy_anything
[ policy_anything ]
commonName              = supplied
emailAddress            = supplied
[ user_cert ]
#SXNetID		= 3:yeak
subjectAltName		= email:copy
basicConstraints	= critical,CA:false
authorityKeyIdentifier	= keyid:always
extendedKeyUsage	= clientAuth,emailProtection
EOT

#  revoke an existing old certificate
if [ -f $CERT.crt ]; then
    openssl ca -revoke $CERT.crt -config ca.config
fi

#  sign the certificate
echo "CA signing: $CERT.csr -> $CERT.crt:"
openssl ca -config ca.config -out $CERT.crt -infiles $CERT.csr
echo "CA verifying: $CERT.crt <-> CA cert"
openssl verify -CAfile ca.crt $CERT.crt

#  cleanup after SSLeay 
rm -f ca.config
rm -f ca.db.serial.old
rm -f ca.db.index.old


sign-server-cert.sh
#!/bin/bash
##
##  sign-server-cert.sh - sign using our root CA the server cert
##  Copyright (c) 2000 Yeak Nai Siew, All Rights Reserved. 
##

HASHALGO="sha256"
VALID_DAYS=730
RANDOM_SRC=/dev/urandom

CN=$1
if [ $# -ne 1 ]; then
    echo "Usage: $0 <www.domain.com>"
    exit 1
fi
if [ ! -f $CN.csr ]; then
    echo "No $CN.csr found. You must create that first."
	exit 1
fi
# Check for root CA key
if [ ! -f ca.key -o ! -f ca.crt ]; then
	echo "You must have root CA key generated first."
	exit 1
fi

# Sign it with our CA key #

#   make sure environment exists
if [ ! -d ca.db.certs ]; then
    mkdir ca.db.certs
fi
if [ ! -f ca.db.serial ]; then
    echo '01' >ca.db.serial
fi
if [ ! -f ca.db.index ]; then
    cp /dev/null ca.db.index
fi


#  create the CA requirement to sign the cert
cat >ca.config <<EOT
[ ca ]
default_ca              = default_CA
[ default_CA ]
dir                     = .
certs                   = \$dir
new_certs_dir           = \$dir/ca.db.certs
database                = \$dir/ca.db.index
serial                  = \$dir/ca.db.serial
RANDFILE                = ${RANDOM_SRC}
certificate             = \$dir/ca.crt
private_key             = \$dir/ca.key
default_days            = ${VALID_DAYS}
default_crl_days        = 30
default_md              = $HASHALGO
preserve                = no
x509_extensions		= server_cert
policy                  = policy_anything
[ policy_anything ]
countryName             = optional
stateOrProvinceName     = optional
localityName            = optional
organizationName        = optional
organizationalUnitName  = optional
commonName              = supplied
emailAddress            = optional
[ server_cert ]
#subjectKeyIdentifier	= hash
authorityKeyIdentifier	= keyid:always
extendedKeyUsage	= serverAuth,clientAuth,msSGC,nsSGC
basicConstraints	= critical,CA:false
[req]
default_md              = $HASHALGO
req_extensions          = v3_req
[ v3_req ]
extendedKeyUsage        = serverAuth, clientAuth
EOT

# Test for Subject Alternate Names
subjaltnames="`openssl req -text -noout -in $CN.csr | sed -e 's/^ *//' | grep -A1 'X509v3 Subject Alternative Name:' | tail -1 | sed -e 's/IP Address:/IP:/g'`"
if [ "$subjaltnames" != "" ]; then
    echo "Found subject alternate names: $subjaltnames"
    echo ""
    echo "subjectAltName          = $subjaltnames" >> ca.config
fi

#  revoke an existing old certificate
if [ -f $CN.crt ]; then
    echo "Revoking current certificate: $CN.crt"
    openssl ca -revoke $CN.crt -config ca.config
fi

#  sign the certificate
echo "CA signing: $CN.csr -> $CN.crt:"
openssl ca -config ca.config -extensions v3_req -out $CN.crt -infiles $CN.csr
echo ""
echo "CA verifying: $CN.crt <-> CA cert"
openssl verify -CAfile ca.crt $CN.crt
echo ""

#  cleanup after SSLeay 
rm -f ca.config
rm -f ca.db.serial.old
rm -f ca.db.index.old


new-server-cert.sh
#!/bin/bash
##
##  new-server-cert.sh - create the server cert
##  Copyright (c) 2000 Yeak Nai Siew, All Rights Reserved. 
##

KEYBITS=2048
HASHALGO="sha256"

# Create the key. This should be done once per cert.
CN=$1
if [ $# -lt 1 ]; then
    echo "Usage: $0 <www.domain.com> [subjectAltName1 [san2 ...]]"
    exit 1
fi

shift
subjectAltNames="$*"

# if private key exists, ask if we want to generate a new key
if [ -f $CN.key ]; then
    read -p "a key for this cn is already existing, generate a new one? " ANSWER
    if [ "$ANSWER" == "Y" ] || [ "$ANSWER" == "y" ]; then
        rm -f $CN.key
    fi
fi

if [ ! -f $CN.key ]; then
    echo "No $CN.key found. Generating one"
    openssl genrsa -out $CN.key $KEYBITS
    echo ""
fi

# Fill the necessary certificate data
CONFIG="server-cert.conf"
cat >$CONFIG <<EOT
[ req ]
default_bits			= $KEYBITS
default_keyfile			= server.key
default_md              = $HASHALGO
distinguished_name		= req_distinguished_name
string_mask			= nombstr
req_extensions			= v3_req
[ req_distinguished_name ]
countryName			= Country Name (2 letter code)
countryName_default		= US
countryName_min			= 2
countryName_max			= 2
stateOrProvinceName		= State or Province Name (full name)
stateOrProvinceName_default	= Texas
localityName			= Locality Name (eg, city)
localityName_default		= Austin
0.organizationName		= Organization Name (eg, company)
0.organizationName_default	= My Personal Organization
organizationalUnitName		= Organizational Unit Name (eg, section)
organizationalUnitName_default	= Secure Server
commonName			= Common Name (eg, www.domain.com)
commonName_default      = $CN
commonName_max			= 64
emailAddress			= Email Address
emailAddress_max		= 40
[ v3_req ]
nsCertType			= server
basicConstraints		= critical,CA:false
keyUsage		= nonRepudiation, digitalSignature, keyEncipherment
EOT

# Handle optional Subject Alternate Names
if [ "$subjectAltNames" != "" ]; then
    echo "subjectAltName                  = @alt_names" >> $CONFIG
    echo "[alt_names]" >> $CONFIG
    numi=1
    numd=1
    cn_already_added=0

    # CN is added to the SAN list automatically
    for san in $CN  $subjectAltNames; do
        # if CN has already been seen, skip it
        if [ "$san" = "$CN" ]; then
            if [ $cn_already_added -eq 0 ]; then
                cn_already_added=1
            else
                continue     #skip to next SAN
            fi
        fi

        # determine if this looks like an IP or a DNS name
        echo $san | egrep '^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$' &> /dev/null
    if [ $? -eq 0 ]; then
            echo "IP.$numi = $san" >> $CONFIG
            let numi++
        else
            echo "DNS.$numd = $san" >> $CONFIG
            let numd++
        fi
    done
fi

echo "Fill in certificate data"
openssl req -new -config $CONFIG -key $CN.key -out $CN.csr

rm -f $CONFIG

echo ""
echo "You may now run ./sign-server-cert.sh to get it signed"
